ТЗ спринта № 9 ver. 1.0

Phronima6 opened the pull request:
Привести в соответствие с ТЗ спринта № 9 ver. 1.0

Phronima6 committed:
Привести в соответствие с ТЗ спринта № 9

Perederey commented on pom.xml:
```diff
@@ -0,0 +1,38 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
вот тут я бы добавил более детальное описание в pom.xml - нет description, url, developers секций. В корпоративной разработке это важно для понимания проектаd

Phronima6 replied:
Спасибо! Все поправлю!

Perederey commented on accounts/pom.xml:
```diff
@@ -0,0 +1,157 @@
+	<modelVersion>4.0.0</modelVersion>
+
+	<parent>
+		<groupId>com.com.bankapp</groupId>
```
у почему тебя тут `com.com.bankapp` Это может вызвать проблемы при деплое и в CI/CD пайплайнах

Perederey commented on docker-compose.yml:
```diff
@@ -0,0 +1,230 @@
+version: '3.8'
```
Отличная реализация Docker Compose!

Perederey commented on accounts/src/main/java/com/bankapp/accounts/service/UserService.java:
```diff
@@ -0,0 +1,257 @@
+    ExchangeFeignClient exchangeFeignClient;
+    ValidationService validationService;
+
+    @Transactional
```
Ну тут конечно лучше переделать метод - он слишком длинный, валидацию стоило бы вынести в отдельный сервис

Perederey commented on gateway/src/main/java/com/bankapp/gateway/CircuitBreakerConfiguration.java:
```diff
@@ -0,0 +1,26 @@
+
+@Configuration
+public class CircuitBreakerConfiguration {
+
```
хорошо тут получилось с Circuit Breaker конфигурацией

Perederey commented on accounts/src/main/java/com/bankapp/accounts/service/UserService.java:
```diff
@@ -0,0 +1,257 @@
+    ValidationService validationService;
+
+    @Transactional
+    public ResponseDto createUser(CreateUserDto createUserDto) {
```
Вот типа такого, псевдо-код, конечно, но примерно так

```java
@Transactional
public ResponseDto createUser(CreateUserDto createUserDto) {
    ResponseDto response = validateCreateUserDto(createUserDto);
    if (response.isHasErrors()) {
        return response;
    }
    
    User user = buildUserFromDto(createUserDto);
    userRepository.save(user);
    return response;
}

private ResponseDto validateCreateUserDto(CreateUserDto dto) {
    ResponseDto response = new ResponseDto();
    
    if (userRepository.findByLogin(dto.getLogin()).isPresent()) {
        response.setHasErrors(true);
        response.getErrors().add("Такой логин уже есть.");
    }
    
    validateAge(dto.getBirthdate(), response);
    validateName(dto.getName(), response);
    validatePasswords(dto.getPassword(), dto.getConfirmPassword(), response);
    
    return response;
}
```

Perederey commented on accounts/src/main/java/com/bankapp/accounts/entity/User.java:
```diff
@@ -0,0 +1,42 @@
+import java.time.LocalDate;
+import java.util.List;
+
+@Entity
```
По хорошему еще посмотри в сторону таких вот штуку, в Spring принято их переопределять:

```java
@Override
public boolean equals(Object o) {
    if (this == o) return true;
    if (o == null || getClass() != o.getClass()) return false;
    User user = (User) o;
    return Objects.equals(login, user.login);
}

@Override
public int hashCode() {
    return Objects.hash(login);
}
```

Perederey commented on front/src/main/java/com/bankapp/front/service/UserService.java:
```diff
@@ -0,0 +1,73 @@
+    AccountsFeignClient accountsFeignClient;
+    UserDetailsServiceImpl userDetailsService;
+
+    public String createUser(CreateUserDto createUserDto, RedirectAttributes model, HttpServletRequest request) {
```
Вообще по-хорошему метод createUser стоило бы разбить на несколько методов - он делает и создание пользователя, и аутентификацию.

Perederey commented on accounts/src/main/java/com/bankapp/accounts/service/NotificationsOutboxService.java:
```diff
@@ -0,0 +1,51 @@
+    }
+
+    @Transactional
+    @Scheduled(fixedDelay = 1000)
```
Ну тут конечно лучше переделать обработку исключений. Мы на работе обычно делаем так, не критично конечно

```java
@Transactional
@Scheduled(fixedDelay = 1000)
public void sendNotification() {
    List<NotificationsOutbox> pendingNotifications = 
        notificationsOutboxRepository.findAllByReceivedFalse();
    
    for (NotificationsOutbox notification : pendingNotifications) {
        try {
            sendSingleNotification(notification);
        } catch (Exception e) {
            log.warn("Failed to send notification with id: {}", notification.getId(), e);
            // Можно добавить retry логику или dead letter queue
        }
    }
}
```,

Perederey commented on exchange/src/main/java/com/bankapp/exchange/service/RateService.java:
```diff
@@ -0,0 +1,68 @@
+    static Map<String, int[]> REALISTIC_RANGES = Map.of(
+        "USD", new int[]{7500, 8500},
+        "CNY", new int[]{1000, 1100}
+    );
```
Map с диапазонами курсов лучше сделать через @ConfigurationProperties и вынести в application.yml

Perederey commented on accounts/src/main/java/com/bankapp/accounts/entity/Account.java:
```diff
@@ -0,0 +1,47 @@
+    @JoinColumn(name = "user_id", nullable = false)
+    User user;
+
+    @Override
```
Хорошо, что переопределил equals и hashCode тут, но ты разве везде это сделал?

Perederey commented on front/src/main/resources/templates/main.html:
```diff
@@ -0,0 +1,215 @@
+<!DOCTYPE html>
```
JavaScript код прямо в HTML не очень хорошо - лучше вынести в отдельные .js файлы

Perederey commented on accounts/src/main/java/com/bankapp/accounts/service/UserService.java:
```diff
@@ -0,0 +1,257 @@
+@AllArgsConstructor
+@FieldDefaults(makeFinal = true, level = AccessLevel.PRIVATE)
+@Service
+public class UserService {
```
Сервис нарушает Single Responsibility Principle - занимается и CRUD операциями, и валидацией, и уведомлениями

Perederey commented on transfer/src/main/java/com/bankapp/transfer/service/TransferService.java:
```diff
@@ -0,0 +1,36 @@
+    ExchangeFeignClient exchangeFeignClient;
+    BlockerFeignClient blockerFeignClient;
+
+    public ResponseDto processTransfer(String login, TransferDto transferDto) {
```
Логика блокировки подозрительных операций смешана с бизнес-логикой - стоило бы вынести в aspect

Perederey commented on exchange/src/main/java/com/bankapp/exchange/service/ExchangeService.java:
```diff
@@ -0,0 +1,30 @@
+@Service
+public class ExchangeService {
+
+    RateRepository rateRepository;
```
и тут кстати, нарушается принцип Dependency Inversion - сервис сильно связан с конкретной реализацией репозитория

Perederey commented on accounts/src/main/java/com/bankapp/accounts/service/NotificationsOutboxService.java:
```diff
@@ -0,0 +1,51 @@
+package com.bankapp.accounts.service;
```
Идеально реализован паттерн Transactional Outbox для надежной доставки уведомлений! 👏

Perederey commented on realm-config/realm-export.json:
```diff
@@ -0,0 +1,2489 @@
+{
```
все нужные клиенты и scope настроены правильно, молодец!

Perederey reviewed (commented):
Привет! Отличная работа с микросервисной архитектурой - видно, что разбираешься в Spring и паттернах корпоративной разработки! Особенно круто реализовал Transactional Outbox pattern и правильно настроил Circuit Breaker.

Из того, на что стоит обратить внимание:
- Некоторые сервисы получились слишком толстыми и нарушают Single Responsibility Principle
- Валидацию стоило бы вынести в отдельные компоненты
- Обработка ошибок в scheduled методах требует доработки

А так молодец! Система получилась масштабируемой и следует многим принципам микросервисной архитектуры!

Commit:
Привести в соответствие с замечаниями:
- исправить groupId в pom.xml файлах;
- рефакторинг UserService (вынести валидацию);
- добавить equals и hashCode во все entity классы;
- вынести конфигурации курсов валют в application.yml;
- вынести JavaScript в отдельные файлы;
- создать Aspect для логики блокировки;
- решить проблему с Dependency Inversion
